org.eclipse.scout.sdk.core.s/src/main/java/org/eclipse/scout/sdk/core/s/permission/PermissionGenerator.java

```diff
@@ -10,33 +10,82 @@
 
-import org.eclipse.scout.sdk.core.java.generator.IJavaElementGenerator;
+import java.util.Optional;
+
+import org.eclipse.scout.sdk.core.java.JavaTypes;
 import org.eclipse.scout.sdk.core.java.generator.field.FieldGenerator;
 import org.eclipse.scout.sdk.core.java.generator.method.IMethodGenerator;
 import org.eclipse.scout.sdk.core.java.generator.method.MethodGenerator;
 import org.eclipse.scout.sdk.core.java.generator.type.PrimaryTypeGenerator;
 import org.eclipse.scout.sdk.core.s.java.apidef.IScoutApi;
+import org.eclipse.scout.sdk.core.s.java.generator.method.ScoutFieldGenerator;
+import org.eclipse.scout.sdk.core.util.Strings;
 
-/**
- * <h3>{@link PermissionGenerator}</h3>
- *
- * @since 5.2.0
- */
-public class PermissionGenerator<TYPE extends PermissionGenerator<TYPE>> extends PrimaryTypeGenerator<TYPE> {
+public class PermissionGenerator extends PrimaryTypeGenerator<PermissionGenerator> {
+
+  public static final String ID_FIELD_NAME = "ID";
+
+  private String m_idRef;
+
+  public static PermissionGenerator create() {
+    return new PermissionGenerator();
+  }
 
   @Override
   protected void setup() {
+    withPreProcessor((e, c) -> {
+      withoutField(field -> field.elementName(c).map(ID_FIELD_NAME::equals).orElse(false));
+      idRef()
+          .filter(ID_FIELD_NAME::equals)
+          .ifPresent(idRef -> withIdRef(null));
+      if (idRef().isEmpty() && c.requireApi(IScoutApi.class).createPermissionIdField()) {
+        withField(ScoutFieldGenerator.createPermissionIdField(elementName().orElseThrow(() -> newFail("Permission name is missing")), ID_FIELD_NAME));
+        withIdRef(ID_FIELD_NAME);
+      }
+    });
     withField(FieldGenerator.createSerialVersionUid())
         .withSuperClassFrom(IScoutApi.class, api -> api.AbstractPermission().fqn())
-        .withMethod(createConstructor(this));
+        .withMethod(createConstructor());
   }
 
-  protected static IMethodGenerator<?, ?> createConstructor(IJavaElementGenerator<?> constructorOwner) {
-    var permissionName = constructorOwner.elementName().orElseThrow(() -> newFail("Permission name is missing"));
+  protected IMethodGenerator<?, ?> createConstructor() {
+    var permissionName = elementName().orElseThrow(() -> newFail("Permission name is missing"));
     return MethodGenerator.create()
         .asPublic()
         .withElementName(permissionName)
-        .withBody(b -> b.superClause().parenthesisOpen().stringLiteral(permissionName).parenthesisClose().semicolon());
+        .withBody(b -> {
+          b.superClause().parenthesisOpen();
+          idRef().ifPresentOrElse(
+              idRef -> {
+                // split into class and field if possible
+                // "ID" -> "ID"
+                // "a.b.c.d.SomeClass.MY_FIELD" -> "a.b.c.d.SomeClass" & "MY_FIELD"
+                var lastDotIndex = idRef.lastIndexOf(JavaTypes.C_DOT);
+                if (lastDotIndex < 0) {
+                  b.ref(idRef);
+                  return;
```

Comment:
this is no reference, just a string to append here? right?
The field name can just be appended with append() like below?